packages/nuxt/src/app/components/layout.ts

```diff
@@ -84,18 +46,66 @@ export default defineComponent({
 
       // We avoid rendering layout transition if there is no layout to render
       return _wrapIf(Transition, hasLayout && transitionProps, {
-        default: () => {
-          const layoutNode = _wrapIf(LayoutLoader, hasLayout && {
+        default: () => h(Suspense, { suspensible: true, onResolve: () => { nextTick(done) } }, {
+          default: () => _wrapIf(LayoutProvider, hasLayout && {
+            layoutProps: mergeProps(context.attrs, { ref: layoutRef }),
             key: layout.value,
             name: layout.value,
-            ...(process.dev ? { hasTransition: !!transitionProps } : {}),
-            ...context.attrs,
-            layoutRef
+            shouldProvide: !props.name,
+            hasTransition: !!transitionProps
           }, context.slots).default()
-
-          return layoutNode
-        }
+        })
       }).default()
     }
   }
 })
+
+const LayoutProvider = defineComponent({
+  name: 'NuxtLayoutProvider',
+  inheritAttrs: false,
+  props: {
+    name: {
+      type: String
+    },
+    layoutProps: {
+      type: Object
+    },
+    hasTransition: {
+      type: Boolean
+    },
+    shouldProvide: {
+      type: Boolean
+    }
+  },
+  setup (props, context) {
+    // Prevent reactivity when the page will be rerendered in a different suspense fork
+    if (props.shouldProvide) {
+      // eslint-disable-next-line vue/no-setup-props-destructure
+      const name = props.name
+      provide('_layout', {
+        isCurrent: (route: RouteLocationNormalizedLoaded) => name === (route.meta.layout ?? 'default')
```

Comment:
Just out of curiosity, what would be the difference between providing the name in plain string over a function?

Reply:
it works just as well, but means we have to move the `?? 'default'` into page.ts, and somehow I felt it made sense for it to be co-located in layout.ts where the matching logic is. But happy to refactor this!